### PR TITLE
[cmd/mdatagen] support map and slice attrs

### PR DESCRIPTION
**Pull request by fredthomsen**
**Description:** <Describe what has changed.>

Add slice and map attributes to resource and datapoint attributes.

**Link to tracking Issue:** <Issue number if applicable>
#18272 

**Testing:** <Describe what testing was performed and which tests were added.>
mdatagen existing tests updated, no new tests added. `make mdatagen-test` passed.

**Documentation:** <Describe the documentation added.>
New example in metadata.yml and auto-generated docs.

### CODE REVIEW

File: cmd/mdatagen/loader.go [hunk 1]
```diff
@@ -89,6 +93,10 @@ func (mvt ValueType) Primitive() string {
 		return "bool"
 	case pcommon.ValueTypeBytes:
 		return "[]byte"
+	case pcommon.ValueTypeSlice:
+		return "[]string"
+	case pcommon.ValueTypeMap:
+		return "map[string]string"
```
Comment (dmitryax) [inline cmd/mdatagen/loader.go:99]: I don't think we should restrict the elements type
```suggestion
	case pcommon.ValueTypeSlice:
		return "[]any"
	case pcommon.ValueTypeMap:
		return "map[string]any"
```

File: cmd/mdatagen/loader.go [hunk 2]
```diff
@@ -65,6 +65,10 @@ func (mvt *ValueType) UnmarshalText(text []byte) error {
 		mvt.ValueType = pcommon.ValueTypeBool
 	case "bytes":
 		mvt.ValueType = pcommon.ValueTypeBytes
+	case "[]string":
+		mvt.ValueType = pcommon.ValueTypeSlice
+	case "map[string]string":
+		mvt.ValueType = pcommon.ValueTypeMap
```
Comment (dmitryax) [inline cmd/mdatagen/loader.go:71]:
```suggestion
	case "slice":
		mvt.ValueType = pcommon.ValueTypeSlice
	case "map":
		mvt.ValueType = pcommon.ValueTypeMap
```

File: cmd/mdatagen/loader.go (hunk 1 above)
Comment (fredthomsen) [reply cmd/mdatagen/loader.go:99]: `any` seems to wide open, which is why I had done it this way.  I originally had it as `any` so I can just revert the second commit on this PR.

File: cmd/mdatagen/loader.go (hunk 2 above)
Comment (fredthomsen) [reply cmd/mdatagen/loader.go]: Yup.  I'll update this as well as the associated data type names in the corresponding example metric yaml file.

**Comment by fredthomsen** [comment]
Realized that I am still linked to my old employer via EasyCLA and I am unable to unlink them.  Need to figure that out before I proceed further.

File: cmd/mdatagen/templates/metrics.go.tmpl [hunk 3]
```diff
@@ -284,6 +288,10 @@ func With{{ $name.Render }}(val {{ $attr.Type.Primitive }}) ResourceMetricsOptio
 		if ras.{{ $name.Render }}.Enabled {
 			{{- if eq $attr.Type.Primitive "[]byte" }}
 			rm.Resource().Attributes().PutEmptyBytes("{{ attributeName $name}}").FromRaw(val)
+            {{- else if eq $attr.Type.Primitive "[]any" }}
+			rm.Resource().Attributes().PutEmptySlice("{{ attributeName $name}}").FromRaw(val)
+            {{- else if eq $attr.Type.Primitive "map[string]any" }}
+			rm.Resource().Attributes().PutEmptyMap("{{ attributeName $name}}").FromRaw(val)
```
Comment (dmitryax) [inline cmd/mdatagen/templates/metrics.go.tmpl:294]:
```suggestion
			{{- else if eq $attr.Type.Primitive "[]any" }}
			rm.Resource().Attributes().PutEmptySlice("{{ attributeName $name}}").FromRaw(val)
			{{- else if eq $attr.Type.Primitive "map[string]any" }}
			rm.Resource().Attributes().PutEmptyMap("{{ attributeName $name}}").FromRaw(val)
```

File: cmd/mdatagen/templates/metrics_test.go.tmpl [hunk 4]
```diff
@@ -114,6 +114,8 @@ func TestMetricsBuilder(t *testing.T) {
 					enabledAttrCount++
 					{{- if $info.Enum }}
 						assert.Equal(t, "{{ index $info.Enum 0 }}", attrVal.Str())
+                    {{- else if or (eq $info.Type.String "Slice") (eq $info.Type.String "Map") }}
```
Comment (dmitryax) [inline cmd/mdatagen/templates/metrics_test.go.tmpl:117]:
```suggestion
					{{- else if or (eq $info.Type.String "Slice") (eq $info.Type.String "Map") }}
```

File: cmd/mdatagen/templates/metrics_test.go.tmpl [hunk 5]
```diff
@@ -160,6 +162,8 @@ func TestMetricsBuilder(t *testing.T) {
 					assert.True(t, ok)
 					{{- if (attributeInfo $attr).Enum }}
 					assert.Equal(t, "{{ index (attributeInfo $attr).Enum 0 }}", attrVal.Str())
+                    {{- else if or (eq (attributeInfo $attr).Type.String "Slice") (eq (attributeInfo $attr).Type.String "Map")}}
```
Comment (dmitryax) [inline cmd/mdatagen/templates/metrics_test.go.tmpl:165]:
```suggestion
					{{- else if or (eq (attributeInfo $attr).Type.String "Slice") (eq (attributeInfo $attr).Type.String "Map")}}
```

**Review by dmitryax — APPROVED** [review]
LGTM, just indentation alignments

**Comment by fredthomsen** [comment]
> LGTM, just indentation alignments

Argh.  my vim setup got all flustered on the template files.

**Review by dmitryax — APPROVED** [review]

**Comment by dmitryax** [comment]
@fredthomsen did you sign the CLA?

**Comment by fredthomsen** [comment]
> @fredthomsen did you sign the CLA?

No.  I am working that out with our company compliance guy now.  I'll message back when I have that resolved.

**Comment by fredthomsen** [comment]
> > @fredthomsen did you sign the CLA?
> 
> No.  I am working that out with our company compliance guy now.  I'll message back when I have that resolved.

Unfortunately I am still waiting on compliance...

**Comment by fredthomsen** [comment]
My organization has signed the CLA; however, I have yet to see that reflected on the EasyCLA portal.

**Comment by fredthomsen** [comment]
Ok CLA issues are resolved. Good to go now.

**Comment by fredthomsen** [comment]
@dmitryax is there anything else you need from me here?

**Review by dmitryax — APPROVED** [review]

**Pull request merged**